exporter/elasticsearchexporter/elasticsearch_bulk.go

```diff
@@ -175,52 +161,138 @@ func createElasticsearchBackoffFunc(config *RetrySettings) func(int) time.Durati
 	}
 }
 
-func shouldRetryEvent(status int) bool {
-	for _, retryable := range retryOnStatus {
-		if status == retryable {
-			return true
+func pushDocuments(ctx context.Context, index string, document []byte, bulkIndexer *esBulkIndexerCurrent) error {
+	return bulkIndexer.Add(ctx, index, bytes.NewReader(document))
+}
+
+func newBulkIndexer(logger *zap.Logger, client *elasticsearch7.Client, config *Config) (*esBulkIndexerCurrent, error) {
+	numWorkers := config.NumWorkers
+	if numWorkers == 0 {
+		numWorkers = runtime.NumCPU()
+	}
+
+	flushInterval := config.Flush.Interval
+	if flushInterval == 0 {
+		flushInterval = 30 * time.Second
+	}
+
+	flushBytes := config.Flush.Bytes
+	if flushBytes == 0 {
+		flushBytes = 5e+6
+	}
+
+	var maxDocRetry int
+	if config.Retry.Enabled {
+		// max_requests includes initial attempt
+		// See https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32344
+		maxDocRetry = config.Retry.MaxRequests - 1
+	}
+	group := &errgroup.Group{}
+	items := make(chan esBulkIndexerItem, config.NumWorkers)
+	stats := bulkIndexerStats{}
+
+	for i := 0; i < numWorkers; i++ {
+		w := worker{
+			indexer:       docappender.NewBulkIndexer(client, 0, maxDocRetry),
+			items:         items,
+			flushInterval: flushInterval,
+			flushTimeout:  config.Timeout,
+			flushBytes:    flushBytes,
+			logger:        logger,
+			stats:         &stats,
 		}
+		group.Go(w.run)
 	}
-	return false
+	return &BulkIndexerPool{
+		items:    items,
+		errgroup: group,
+		stats:    &stats,
+	}, nil
+}
+
+type bulkIndexerStats struct {
+	docsIndexed atomic.Int64
+}
+
+type BulkIndexerPool struct {
```

Comment:
What do you think about designing the `BulkIndexerPool` with a channel containing `NumWorkers` count of bulk indexers? For adding we can take a bulk indexer from the channel and add stuff to it. Once done we can decide to flush and put it back in the pool. Another option for flush could be to just flush the indexers one by one periodically in a single goroutine defined on the `BulkIndexerPool` itself.

Reply:
I have thought about this. While it'll be nice to have this enhancement, the current BulkIndexerPool is the closest to the go-elasticsearch BulkIndexer implementation, so it will make this PR easier to review. Also, with https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32377 the flush behavior and mechanism will be vastly different. I'm not sure if it is worth it to fix the sparse bulk request behavior now.